Proper Int64

mitschabaude opened the pull request:
* Replace unfinished, unconstrained Int64 with a proper implementation
* Closes #66 
* Make Int64 API less boilerplate-y by allowing `number`, `string`, etc as method arguments
* Remove some hacks in toJson conversion, possible bc the new Int64 has the same layout as `Currency.Amount.Signed.t`

This is stacked on top of https://github.com/o1-labs/snarkyjs/pull/164, so that it can:
* Make previously failing Int64 unit tests work
* Add more Int64 unit tests

Since this will use updated bindings from https://github.com/MinaProtocol/mina/pull/11042, which fix an aspect of the `Field` constructor, this also adds a few new tests for the Field constructor.
* Closes #43

mitschabaude commented on src/lib/int.ts:
```diff
@@ -301,119 +299,183 @@ export class UInt32 extends CircuitValue {
-    return Int64.sizeInFields();
+function argToField(
+  name: string,
+  x: { value: Field } | number | string
```
TODO: add range checks (not constraints, just checks) on constant UInt64 / UInt32 / Int64 as you create them

mitschabaude replied:
TODO: add range checks (not constraints, just checks) on constant UInt64 / UInt32 / Int64 as you create them

mimoo commented on src/lib/int.ts:
```diff
@@ -301,119 +299,183 @@ export class UInt32 extends CircuitValue {
+  if (typeof x === 'bigint') return Int64.fromBigInt(x);
+  if (x instanceof UInt64 || x instanceof UInt32) return Int64.fromUnsigned(x);
+  return x;
+}
```
why is this not implemented on `Int64`, something like `Int64.from(x: number | string | bigint | UInt64 | UInt32 | Int64)`

and same issue here, the `fromNumber` and other similar functions don't error if the user tries to input values that are too large (or too small)

mitschabaude replied:
great suggestion!

mimoo commented on src/lib/int.ts:
```diff
@@ -53,7 +42,7 @@ export class UInt64 extends CircuitValue {
   static NUM_BITS = 64;
 
-  divMod(y: UInt64 | number): [UInt64, UInt64] {
+  divMod(y: UInt64 | number | string): [UInt64, UInt64] {
```
this return type is a bit error prone, why not return something like `{ quotient, rest }` instead?

Especially when we do things like `let  [q] =` or `let [, r] = `

mimoo commented on src/lib/int.ts:
```diff
@@ -301,119 +299,183 @@ export class UInt32 extends CircuitValue {
-    throw 'repr';
+  equals(y: Int64 | number | string | bigint | UInt64 | UInt32) {
+    let y_ = argToInt64(y);
+    return this.toField().equals(y_.toField());
```
would `this.magnitude.equals(y_.magnitude).and(this.sign.equals(y_.sign)` be better? The two equalities could be optimized by the compiler (using the permutation), and thus you'd only have one constraint (the `and`)

mitschabaude replied:
I didn't do that because I was worried about special-casing for 0, where -0 should equal 0.. The special-casing would also require multiplication constraints I think

mimoo replied:
good point :o

mimoo reviewed (approved):
This is awesome! Code is clear, and this looks good to me, left a few minor comments

mitschabaude commented on src/lib/int.ts:
```diff
@@ -301,119 +299,183 @@ export class UInt32 extends CircuitValue {
-  static check(x: Sgn) {
-    let x_ = x.value.seal();
-    x_.mul(x_).assertEquals(Field.one);
+  static fromFieldUnchecked(x: Field) {
```
good point, I'll add `private`, that makes it not visible to typescript (even though it's on the class)

mitschabaude commented on src/lib/int.ts:
```diff
@@ -301,119 +299,183 @@ export class UInt32 extends CircuitValue {
-    return new Sgn(Field.one);
+  static fromNumber(x: number) {
+    if (x < 0) return Int64.fromString(String(-x)).neg();
+    return Int64.fromString(String(x));
```
no, it also has `.neg()` at the end

mimoo replied:
🤦🏻‍♂️

mitschabaude commented on src/lib/int.ts:
```diff
@@ -301,119 +299,183 @@ export class UInt32 extends CircuitValue {
-  */
+  // --- circuit-compatible operations below ---
+  // the assumption here is that all Int64 values that appear in a circuit are already checked as valid
+  // this is because Circuit.witness calls .check
```
https://github.com/MinaProtocol/mina/blob/39af86e4a9f802277cdf73a6842c43fc4da48546/src/lib/snarky_js_bindings/lib/snarky_js_bindings_lib.ml#L1365

mimoo replied:
🙏🏻

mitschabaude commented on src/lib/int.ts:
```diff
@@ -301,119 +299,183 @@ export class UInt32 extends CircuitValue {
-  static fromUnsigned(x: UInt64): Int64 {
-    return new Int64(x.value);
+  static get minusOne() {
+    return new Int64(UInt64.one, Field.one.neg());
```
I like it! I'll make `Field.one` the default argument for the sign

mitschabaude commented on src/lib/int.ts:
```diff
@@ -301,119 +299,183 @@ export class UInt32 extends CircuitValue {
-  neg(): Int64 {
-    return new Int64(this.value.neg());
+  add(y: Int64 | number | string | bigint | UInt64 | UInt32) {
+    let y_ = argToInt64(y);
```
It actually doesn't, when it goes through `fromFieldUnchecked` the range is checked.. I'll make sure that's done in all cases and add comments about it

mitschabaude commented on src/lib/int.ts:
```diff
@@ -301,119 +299,183 @@ export class UInt32 extends CircuitValue {
-    return new Int64(this.value.neg());
+  add(y: Int64 | number | string | bigint | UInt64 | UInt32) {
+    let y_ = argToInt64(y);
+    return Int64.fromField(this.toField().add(y_.toField()));
```
Yeah I think you may have a point.. I might reconsider this when range checks become cheap.

Do you mean shifting with `2^64 + res` and range-checking that? I don't follow how it works with multiplication & negative numbers.

Currently, it's convenient to have Int64 be {magnitude, sign} because that mirrors the Ocaml data structure. If I make it a single Field constrained to 64 bits, I allow a slightly smaller range than the Ocaml type (64 vs 65 bits)

mimoo replied:
yeah `+` not `*` sorry

mitschabaude commented:
@mimoo addressed all feedback,
also added Field.to/fromBigint() to make the code cleaner in some places

this should be good to go, will merge once the parent PR is merged
